src/RowWiseSparseAdagradFused.cc

```diff
@@ -123,10 +138,10 @@ typename ReturnFunctionSignature<indxType, offsetType>::
         x86::Gp lengths = a->gpz(11);
         x86::Xmm epsilon(0);
         x86::Xmm lr(1);
-        x86::Gp mask_avx2 = a->gpz(12);
+        x86::Gp rand_buffer = a->gpz(12);
 
-        // reuse mask_avx2 because mask_avx2 is used only at the beginning
-        x86::Gpd lengths_R = a->gpz(12).r32();
+        // FP32 weights does not need rand_buffer
+        x86::Gpd lengths_R = areWeightsFp16 ? a->zbx().r32() : a->gpz(12).r32();
```

Comment:
Maybe to simplify we always use r12 for lengths_R and rand_buffer can use zax? zax is a caller-saved register like zcx and zdx, so no need to set dirty flag separately for areWeightsFp16 case as in line 184.